examples/seismic/source.py

```diff
@@ -1,60 +1,144 @@
+from scipy import signal
+
 from devito import Dimension
 from devito.function import SparseTimeFunction
 from devito.logger import error
 
+from cached_property import cached_property
+
+from copy import copy
 import numpy as np
 try:
     import matplotlib.pyplot as plt
 except:
     plt = None
 
-__all__ = ['PointSource', 'Receiver', 'Shot', 'RickerSource', 'GaborSource']
+__all__ = ['PointSource', 'Receiver', 'Shot', 'WaveletSource',
+           'RickerSource', 'GaborSource', 'TimeAxis']
+
+
+class TimeAxis:
+    """ Data object to store the time axis. Exactly three of the four key arguments
+        must be prescribed. Because of remainder values it is not possible to create
+        a time axis that exactly adhears to the inputs therefore start, stop, step
+        and num values should be taken from the TimeAxis object rather than relying
+        upon the input values.
+
+        The four possible cases are:
+        start is None: start = step*(1 - num) + stop
+        step is None: step = (stop - start)/(num - 1)
+        num is None: num = ceil((stop - start + step)/step);
+                     because of remainder stop = step*(num - 1) + start
+        stop is None: stop = step*(num - 1) + start
+
+    :param start:(Optional) Start of time axis.
+    :param step: (Optional) Time interval.
+    :param: num: (Optional) Number of values (Note: this is the number of intervals + 1).
+                 stop value is reset to correct for remainder.
+    :param stop: (Optional) End time.
+    """
+    def __init__(self, start=None, step=None, num=None, stop=None):
+        try:
+            if start is None:
+                start = step*(1 - num) + stop
+            elif step is None:
+                step = (stop - start)/(num - 1)
+            elif num is None:
+                num = int(np.ceil((stop - start + step)/step))
+                stop = step*(num - 1) + start
+            elif stop is None:
```

Comment:
Shouldn't these be all seperate `if`. In the case `start` is None, the rest is never checked